todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/local/TasksLocalDataSource.kt

```diff
@@ -0,0 +1,131 @@
+    checkNotNull(context, "context cannot be null")
+    checkNotNull(schedulerProvider, "scheduleProvider cannot be null")
+    val dbHelper = TasksDbHelper(context)
+    val sqlBrite = SqlBrite.Builder().build()
+    databaseHelper = sqlBrite.wrapDatabaseHelper(dbHelper, schedulerProvider.io())
+    taskMapperFunction = { this.getTask(it) }
+  }
+
+  override fun getTasks(): Flowable<List<Task>> {
+    val projection = listOf(
+        TaskEntry.COLUMN_NAME_ENTRY_ID,
+        TaskEntry.COLUMN_NAME_TITLE,
+        TaskEntry.COLUMN_NAME_DESCRIPTION,
+        TaskEntry.COLUMN_NAME_COMPLETED
+    )
+    val sql = "SELECT ${TextUtils.join(",", projection)} FROM ${TaskEntry.TABLE_NAME}"
+    return databaseHelper
+        .createQuery(TaskEntry.TABLE_NAME, sql)
+        .mapToList(taskMapperFunction)
+        .toFlowable(BackpressureStrategy.BUFFER)
+  }
+
+  private fun getTask(c: Cursor): Task = with(c) {
+    val itemId = getString(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_ENTRY_ID))
+    val title = getString(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_TITLE))
+    val description = getString(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_DESCRIPTION))
+    val completed = getInt(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_COMPLETED)) == 1
+    val details = TaskDetails.builder().title(title).description(description).completed(completed).build()
+    return@with Task.create(itemId, details)
+  }
+
+  override fun getTask(taskId: String): Flowable<Optional<Task>> {
+    val projection = listOf(
+        TaskEntry.COLUMN_NAME_ENTRY_ID,
+        TaskEntry.COLUMN_NAME_TITLE,
+        TaskEntry.COLUMN_NAME_DESCRIPTION,
+        TaskEntry.COLUMN_NAME_COMPLETED
+    )
+    val sql =
+            "SELECT ${TextUtils.join(",", projection)} " +
+            "FROM ${TaskEntry.TABLE_NAME} " +
+            "WHERE ${TaskEntry.COLUMN_NAME_ENTRY_ID} LIKE ?"
+    return databaseHelper
+        .createQuery(TaskEntry.TABLE_NAME, sql, taskId)
+        .mapToOneOrDefault(
+            { Optional.of(taskMapperFunction(it)) }, Optional.absent<Task>())
+        .toFlowable(BackpressureStrategy.BUFFER)
+  }
+
+  override fun saveTask(task: Task) {
+    checkNotNull(task)
+    val values = ContentValues().apply {
+      put(TaskEntry.COLUMN_NAME_ENTRY_ID, task.id)
+      put(TaskEntry.COLUMN_NAME_TITLE, task.details.title)
+      put(TaskEntry.COLUMN_NAME_DESCRIPTION, task.details.description)
+      put(TaskEntry.COLUMN_NAME_COMPLETED, task.details.completed)
+    }
+    databaseHelper.insert(TaskEntry.TABLE_NAME, values, SQLiteDatabase.CONFLICT_REPLACE)
+  }
+
+  override fun deleteAllTasks() {
+    databaseHelper.delete(TaskEntry.TABLE_NAME, null)
+  }
+
+  override fun deleteTask(taskId: String) {
+    val selection = TaskEntry.COLUMN_NAME_ENTRY_ID + " LIKE ?"
+    val selectionArgs = arrayOf(taskId)
+    databaseHelper.delete(TaskEntry.TABLE_NAME, selection, *selectionArgs)
+  }
+
+  companion object {
+    private var INSTANCE: TasksLocalDataSource? = null
```

Comment:
I don't think we should use this pattern. I know it was used in the old java version, but why don't we just create this class somewhere and inject it where needed instead?